belt-dwp/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+        let cipher: BeltCtr = BeltCtr::new(&key, nonce);
+
+        let _s = to_u32::<4>(nonce);
+        let _k = to_u32::<8>(&key);
+        // 2.1. 𝑠 ← belt-block(𝑆, 𝐾);
+        let s = belt_block_raw(_s, &_k);
+        // 2.2. 𝑟 ← belt-block(𝑠, 𝐾);
+        let r = from_u32::<16>(&belt_block_raw(s, &_k));
+
+        Self {
+            enc_cipher: cipher,
+            mac_cipher: BeltBlock::new(&key),
+            // Unwrap is safe because the key is always 16 bytes
+            ghash: GHash::new_with_init_block(&Key::<GHash>::try_from(&r[..]).unwrap(), T),
+        }
+    }
+
+    fn encrypt_in_place_detached(
+        &mut self,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+    ) -> aead::Result<Tag> {
+        let sizes_block =
+            get_sizes_block(associated_data.len() as u64 * 8, buffer.len() as u64 * 8);
+
+        // 3. For 𝑖 = 1, 2, . . . , 𝑚 do:
+        //  3.1 𝑡 ← 𝑡 ⊕ (𝐼𝑖 ‖ 0^{128−|𝐼𝑖|})
+        //  3.2 𝑡 ← 𝑡 * 𝑟.
+        self.ghash.update_padded(associated_data);
+
+        // 4. For 𝑖 = 1, 2, . . . , 𝑛 do:
+        //  4.1 𝑠 ← 𝑠 ⊞ ⟨1⟩_128
+        //  4.2 𝑌𝑖 ← 𝑋𝑖 ⊕ Lo(belt-block(𝑠, 𝐾), |𝑋𝑖|)
+        //  4.3 𝑡 ← 𝑡 ⊕ (𝑌𝑖 ‖ 0^{128−|𝑌𝑖|})
+        //  4.4 𝑡 ← 𝑡 * 𝑟.
+        buffer.chunks_mut(16).for_each(|block| {
+            self.enc_cipher.apply_keystream(block);
+            self.ghash.update_padded(block);
+        });
+
+        // 5. 𝑡 ← 𝑡 ⊕ (⟨|𝐼|⟩_64 ‖ ⟨|𝑋|⟩_64)
+        self.ghash.xor_s(&sizes_block);
+
+        // 6. 𝑡 ← belt-block(𝑡 * 𝑟, 𝐾).
+        let mut tag = self.finish_tag();
+
+        self.mac_cipher.encrypt_block(&mut tag);
+
+        // Unwrap is safe because the tag is always 8 bytes
+        Ok(Tag::try_from(&tag[..8]).unwrap())
+    }
+
+    fn decrypt_in_place_detached(
+        &mut self,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+        tag: &Tag,
+    ) -> aead::Result<()> {
+        let sizes_block =
+            get_sizes_block(associated_data.len() as u64 * 8, buffer.len() as u64 * 8);
+
+        // 3. For 𝑖 = 1, 2, . . . , 𝑚 do:
+        //  3.1 𝑡 ← 𝑡 ⊕ (𝐼𝑖 ‖ 0^{128−|𝐼𝑖|})
+        //  3.2 𝑡 ← 𝑡 * 𝑟.
+        self.ghash.update_padded(associated_data);
+        // 4. For 𝑖 = 1, 2, . . . , 𝑛 do:
+        //  4.1 𝑡 ← 𝑡 ⊕ (𝑌𝑖 ‖ 0^{128−|𝑌𝑖|})
+        //  4.2 𝑡 ← 𝑡 * 𝑟.
+        self.ghash.update_padded(buffer);
+
+        // 5. 𝑡 ← 𝑡 ⊕ (⟨|𝐼|⟩_64 ‖ ⟨|𝑋|⟩_64)
+        self.ghash.xor_s(&sizes_block);
+
+        let mut tag_exact = self.finish_tag();
+
+        // 6. 𝑡 ← belt-block(𝑡 * 𝑟, 𝐾).
+        self.mac_cipher.encrypt_block(&mut tag_exact);
+
+        // 7. If 𝑇 != Lo(𝑡, 64), return ⊥
+        if tag_exact[..8] != tag[..] {
+            return Err(aead::Error);
+        }
+
+        // 8. For 𝑖 = 1,2,...,𝑛 do:
+        // 8.1. 𝑠 ← 𝑠 ⊞ ⟨1⟩128;
+        // 8.2. 𝑋𝑖 ← 𝑌𝑖 ⊕ Lo(belt-block(𝑠, 𝐾), |𝑌𝑖|)
+        buffer.chunks_mut(16).for_each(|block| {
+            self.enc_cipher.apply_keystream(block);
+        });
```

Comment:
You don't need to chunk data here. You can simply call `apply_keystream` on `buffer`. Also note that one-pass processing could be more efficient (i.e. you would update the `ghash` state and decrypt on each block), but in the case of tag mismatch you would need to either re-encrypt or zeroize output buffer (see #660).

Reply:
Thank you! Done.
Looks like i don't need to re-encrypt, because encryption is only applied after tag calculating?
It is not so optimized, but looks fine, as I suggest.

Reply:
Yes, with two-pass decryption you don't need to re-encrypt. We can keep the current implementation and leave one-pass implementation for a potential later PR.